lib/manageiq/providers/ibm_cloud/cloud_tools/authentication.rb

```diff
@@ -61,27 +61,37 @@ class BearerAuth < IbmVpc::Authenticators::BearerTokenAuthenticator
             # @return [void]
             def initialize(bearer_info, logger: nil)
               @logger = define_logger(logger)
-              @bearer_info = verify_info(bearer_info)
+              @bearer_info = bearer_info
               super({:bearer_token => bearer_info[:token]})
             end
 
             # @return [Hash{Symbol=>String, Integer}] bearer hash with token, expire_time and api_key as keys.
             attr_reader :bearer_info
 
+            def authenticate(headers)
+              verify_bearer
+              super(headers)
+            end
+
             private
 
-            # Verify that the bearer token hasn't expired. If it has and an API Key is present try to get a new bearer token.
-            #
-            # @return [Hash{Symbol => String, Integer}] @see #bearer_info
-            def verify_info(bearer_info)
-              # Raise standard error if expiration time expires in the next 10 second.
-              if expired?(bearer_info[:expire_time])
-                raise 'Bearer token has expired.' if bearer_info[:api_key].nil?
-
-                @logger.info('Bearer token expired. Fetching new one using API Key.')
-                return IamAuth.new(bearer_info[:api_key]).bearer_info
+            # Check to see if bearer token has expired. If it has fetch a new one.
+            # @return [void]
+            def verify_bearer
+              if expired?(@bearer_info[:expire_time])
+                @bearer_info = new_bearer(@bearer_info)
+                @bearer_token = @bearer_info[:token]
               end
-              bearer_info
+            end
+
+            # Get a new bearer token.
+            # @param bearer_info [Hash{Symbol=>String, Integer}] Bearer info hash from IamAuth
+            # @return [Hash{Symbol=>String, Integer}] Bearer info hash from IamAuth
+            def new_bearer(bearer_info)
+              raise 'Bearer token has expired.' if bearer_info[:api_key].nil?
```

Comment:
A little confusing that the exception message here is the token has expired, but the log message [here](https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/189/files#diff-64f4a90e4d7c883bcf2208d05e76761872252560a972a7f4bfc46c084cf2e7baR93) is that it has expired.

Would it be more accurate to say `Cannot refresh bearer token without an API key` or something to that effect?

Reply:
Is not providing an api_key a valid scenario?  Could we just ensure that the api_key is always passed in or are there some places where we have to only provide a bearer token?

Reply:
Updating message to say 'Bearer token has expired and unable to refresh. An api key is not present in bearer_info hash.'

The api key should be automatically included when fetching the bearer_info via standard means. However it is just a hash and there may be times when that hash doesn't have the api key. I don't want to validate the hash because all that is necessary to work is a valid bearer_token.  If it is a short-lived operation this may be all that is necessary. The presence of the bearer_token is validated on init by super. The presence of expire_time is validated in `expired?`. Finally the presence of api_key is validated during `new_bearer`.

Reply:
Actually, my explanation about the short-lived is no longer true. All the keys are validated just not at once. This change makes it impossible to send a request unless all the keys are present and valid.

Reply:
If we wanted to we could return false on expire_time.nil? for expired? . This would allow for a simple bearer_token to be passed in. If it is expired then an 401 will be raised.